fenix/app/src/main/java/org/mozilla/fenix/telemetry/TelemetryMiddleware.kt

```diff
@@ -54,12 +64,31 @@ class TelemetryMiddleware(
         // Pre process actions
 
         when (action) {
+            is ContentAction.UpdateIsSearchAction -> {
+                if (action.isSearch && nimbusSearchEngine.isNotEmpty() &&
+                    (action.searchEngineName == nimbusSearchEngine)
+                ) {
+                    val timerId = timerIdFunc.invoke()
+                    searchState[action.sessionId] = timerId
+                }
+            }
             is ContentAction.UpdateLoadingStateAction -> {
                 context.state.findTab(action.sessionId)?.let { tab ->
+                    val hasFinishedLoading = tab.content.loading && !action.loading
+
                     // Record UriOpened event when a non-private page finishes loading
-                    if (tab.content.loading && !action.loading) {
+                    if (hasFinishedLoading) {
                         Events.normalAndPrivateUriCount.add()
                     }
+
+                    // We'll ignore any type of consent page as this will interfere with loading times.
+                    val progressCompleted = tab.content.progress == PROGRESS_COMPLETE
+                    if (hasFinishedLoading && progressCompleted) {
+                        searchState[action.sessionId]?.let {
+                            Metrics.searchPageLoadTime.stopAndAccumulate(it)
+                        }
+                        searchState.remove(action.sessionId)
```

Comment:
Should we remove the entry from `searchState` when we've finished loading even if progress hasn't been completed?

Reply:
I'm thinking probably not, we could get a situation where loading has finished, but progress hasn't completed, then when it has completed it would no longer exist in the map to stop?

Reply:
@t-p-white In that case we still wouldn't be able to send the telemetry as we only run this code when the loading state changes. Maybe we should create a `tryToFinish` function that we call here as well as when progress changes on the content state

Reply:
@boek good callout, thanks! I think given the use case, we can just simplify to just cancel the timer when progress is not completed